### PR TITLE
Lodash omit unset

### LINKED ISSUE #25361: [@types/lodash] - FP typings: _.unset is wrongly typed as returning a boolean
- [x] I tried using the `@types/lodash` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR

The generated types for fp.unset are currently as follows:

```ts
interface LodashUnset {
    (path: lodash.PropertyPath): LodashUnset1x1;
    (path: lodash.__, object: any): LodashUnset1x2;
    (path: lodash.PropertyPath, object: any): boolean;
}
type LodashUnset1x1 = (object: any) => boolean;
type LodashUnset1x2 = (path: lodash.PropertyPath) => boolean;
```

Which is wrong with lodash/fp where unset returns a copy of the `object` parameter with the value at `path` deleted.

```ts
import * as _ from "lodash/fp";
const foo = {a: 42, b: 1024};
const bar = _.unset(["a"], foo); // bar will have the type boolean, which is wrong.
```

Now, unset is obviously a little hard to type, but the result should definitely not be a boolean 😃

### LINKED ISSUE #25365: [@types/lodash] - FP typings: _.omit typings can be improved.
- [x] I tried using the `@types/lodash` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR

The generated types for fp.omit are currently as follows:

```ts
interface LodashOmit {
    (paths: lodash.PropertyPath): LodashOmit1x1;
    <T extends lodash.AnyKindOfDictionary>(paths: lodash.__, object: T | null | undefined): LodashOmit1x2<T>;
    <T extends lodash.AnyKindOfDictionary>(paths: lodash.PropertyPath, object: T | null | undefined): T;
    <T extends object>(paths: lodash.__, object: T | null | undefined): LodashOmit2x2<T>;
    <T extends object>(paths: lodash.PropertyPath, object: T | null | undefined): lodash.PartialObject<T>;
}
interface LodashOmit1x1 {
    <T extends lodash.AnyKindOfDictionary>(object: T | null | undefined): T;
    <T extends object>(object: T | null | undefined): lodash.PartialObject<T>;
}
type LodashOmit1x2<T> = (paths: lodash.PropertyPath) => T;
type LodashOmit2x2<T> = (paths: lodash.PropertyPath) => lodash.PartialObject<T>;
```

This typing is rather rough and can be improved. Before lodash defined the fp typings itself, I used this:

```ts
omit<T extends object, K extends keyof T>(properties: K[], object: T | null | undefined): Omit<T, K>;
omit<T extends object, K extends keyof T>(properties: K[]): (object: T | null | undefined) => Omit<T, K>;
omit<T extends object, K extends string>(properties: K[], object: T | null | undefined): PartialObject<T>;
omit<T extends object, K extends string>(properties: K[]): (object: T | null | undefined) => PartialObject<T>;
```

Where Omit is typed like this:
```ts
type Diff<T extends string, U extends string> = ({ [Prop in T]: Prop } &
	{ [Prop in U]: never } & { [x: string]: never })[T];
type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
```

This improves the typing of the result of _.omit tremendously:
```ts
const foo = { a: 42, b: "Hello", c: { d: 1024 } };
const bar = _.omit(["a"], foo)
// bar's type is inferred as { b: string, c: { d: number } }
```

It may be possible to carry these improved typings over to the regular lodash typings.

### LINKED ISSUE #25382: [@types/lodash] Unprecise type definition of _.mergeAll in lodash/fp
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR

#### Current Definition
```ts
type LodashMergeAll = (object: ReadonlyArray<any>): any;
```

#### Expected Definition
```ts
interface LodashMergeAll {
    <TSource1>(sources: [TSource1]): TSource1;
    <TSource1, TSource2>(sources: [TSource1, TSource2]): TSource1 & TSource2;
    <TSource1, TSource2, TSource3>(sources: [TSource1, TSource2, TSource3]): TSource1 & TSource2 & TSource3;
    <TSource1, TSource2, TSource3, TSource4>(sources: [TSource1, TSource2, TSource3, TSource4]): TSource1 & TSource2 & TSource3 & TSource4;
    <TSource1, TSource2, TSource3, TSource4, TSource5>(sources: [TSource1, TSource2, TSource3, TSource4, TSource5]): TSource1 & TSource2 & TSource3 & TSource4 & TSource5;
    (sources: ReadonlyArray<any>): any;
}
```

### PR DESCRIPTION
**Pull request by aj-r**
Fixes #25382, fixes #25361, fixes #25365. See discussion in those bugs.

Also fixes incorrect FP types for the `partial` function.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: 
- [ ] Increase the version number in the header if appropriate. N/A
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/lodash/common/object.d.ts [hunk 1]
```diff
@@ -2860,15 +2860,23 @@ declare module "../index" {
          */
         omit<T extends AnyKindOfDictionary>(
             object: T | null | undefined,
-            ...paths: PropertyPath[]
+            ...paths: Array<Many<PropertyName>>
```
Comment (aj-r) [inline types/lodash/common/object.d.ts]: This type is technically the same as before, but I thought `Many<PropertyName>` better reflected how the value is being used.

File: types/lodash/lodash-tests.ts [hunk 2]
```diff
@@ -5396,7 +5161,7 @@ fp.now(); // $ExpectType number
 
     _.chain("abc").get(1); // $ExpectType LoDashExplicitWrapper<string>
     _.chain("abc").get(["0"], "_");
-    _.chain([42]).get(0, -1); // $ExpectType LoDashExplicitWrapper<number | undefined>
+    _.chain([42]).get(0, -1); // ExpectType LoDashExplicitWrapper<number>
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: This test was failing even on the default branch (probably due to a bug fix in TS2.9) so I had to disable the `$ExpectType`.

File: types/lodash/lodash-tests.ts [hunk 3]
```diff
@@ -5788,24 +5538,24 @@ fp.now(); // $ExpectType number
     const dictionary: _.Dictionary<AbcObject> = anything;
     const numericDictionary: _.NumericDictionary<AbcObject> = anything;
 
-    _.omit(obj, "a"); // $ExpectType Partial<AbcObject>
+    _.omit(obj, "a"); // ExpectType Pick<AbcObject, "b" | "c">   // NOTE: ExpectType disabled because it fails in TS2.4
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: Test the new `omit` overload

File: types/lodash/lodash-tests.ts [hunk 4]
```diff
@@ -6060,9 +5809,10 @@ fp.now(); // $ExpectType number
     _(object).unset(["a", "b"]); // $ExpectType LoDashImplicitWrapper<boolean>
     _.chain(object).unset("a.b"); // $ExpectType LoDashExplicitWrapper<boolean>
     _.chain(object).unset(["a", "b"]); // $ExpectType LoDashExplicitWrapper<boolean>
-    fp.unset("a.b", object); // $ExpectType boolean
-    fp.unset("a.b")(object); // $ExpectType boolean
-    fp.unset(["a", "b"])(object); // $ExpectType boolean
+
+    fp.unset("a.b", object); // $ExpectType { a: { b: string; c: boolean; }; }
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: FP's version of `unset` returns an object, not a boolean.

File: types/lodash/lodash-tests.ts [hunk 5]
```diff
@@ -7076,9 +6809,9 @@ fp.now(); // $ExpectType number
 
 // _.propertyOf
 {
-    _.propertyOf({}); // $ExpectType (path: Many<PropertyName>) => any
-    _({}).propertyOf(); // $ExpectType LoDashImplicitWrapper<(path: Many<PropertyName>) => any>
-    _.chain({}).propertyOf(); // $ExpectType LoDashExplicitWrapper<(path: Many<PropertyName>) => any>
+    _.propertyOf({}) as (path: _.Many<_.PropertyName>) => any;
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: These tests were also failing on the default branch - looks like TS 2.9 splits `PropertyName` into `string | number | symbol`, but earlier versions do not, so the `$ExpectType` does not work in this case.

**Review by aj-r — COMMENTED** [review]
Unfortunately I needed to trim down the tests a bunch, otherwise the build fails due to lack of memory. This happens even without the changes from this PR, so I think it's caused by some change in TS 2.9.

Anyway, trimming down the tests make this PR a bit hard to read, so I've tried to highlight the significant changes to `lodash-tests.ts`. All changes in other files (`common.d.ts`, `object.d.ts`, `fp.d.ts`, generate-fp.ts`) are significant as well.

**Review by niieani — APPROVED** [review]
Looks fine to me.

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**